### PR TITLE
ICOS CO2 Reader

### LINKED ISSUE #862: Add ICOS CO2 data
We have been asked to add ICOS surface concentrations of CO2 into CAMS2_82. The data use the HARP-1.0 conventions, so we will likely be able to adapt/modify the MEP reader to read them in, but new a reader is expected to be needed. 

Example file:
```
netcdf icos-co2-nrt-zsf-3.0m-20230401-20230403T100002 {
dimensions:
	string_3 = 3 ;
	string_9 = 9 ;
	string_176 = 176 ;
	string_24 = 24 ;
	string_53 = 53 ;
	time = 24 ;
variables:
	char site_code(string_3) ;
	char site_name(string_9) ;
	char citation(string_176) ;
	char doi(string_24) ;
	char landing_page_url(string_53) ;
	double datetime_start(time) ;
		datetime_start:units = "days since 2000-01-01" ;
	double datetime_stop(time) ;
		datetime_stop:units = "days since 2000-01-01" ;
	float latitude ;
		latitude:units = "degree_north" ;
	float longitude ;
		longitude:units = "degree_east" ;
	float surface_altitude ;
		surface_altitude:units = "m" ;
	float altitude ;
		altitude:units = "m" ;
	double CO2_volume_mixing_ratio(time) ;
		CO2_volume_mixing_ratio:units = "ppm" ;

// global attributes:
		:Conventions = "HARP-1.0" ;
		:datetime_start = 8491. ;
		:datetime_stop = 8492. ;
		:source_product = "ICOS_ATC_NRT_ZSF_2022-03-01_2023-04-01_3.0_880-1167.CO2" ;
		:history = "2023-04-03T08:00:02Z harpicos \'ICOS_ATC_NRT_ZSF_2022-03-01_2023-04-01_3.0_880-1167.CO2\'" ;
}
```

Will add the [pyaerocom-preproc](https://github.com/metno/pyaerocom-preproc) results once ready.

### PR DESCRIPTION
**Pull request by lewisblake**
Designed to close #862

### CODE REVIEW

**Comment by lewisblake** [comment]
For this PR we need to
- [ ] Download the data, aggregate it to yearly files (e.g., using cdo), place the aggreagetd files in `lustre/storeB/project/aerocom/aerocom1/AEROCOM_OBSDATA/ICSO/aggregated/`
- [ ] Choose sample files (I have some locally) for testing, write tests, and place test files on the server
- [ ] Run a few Aeroval experiments as tests (obs only, with IFS output)

**Comment by lewisblake** [comment]
(ICOS Carbon Portal SPARQL Client)[https://meta.icos-cp.eu/sparqlclient/?type=CSV] query to get CO2 data

```SQL
prefix cpmeta: <http://meta.icos-cp.eu/ontologies/cpmeta/>
prefix prov: <http://www.w3.org/ns/prov#>
prefix xsd: <http://www.w3.org/2001/XMLSchema#>
select ?dobj ?hasNextVersion ?spec ?fileName ?size ?submTime ?timeStart ?timeEnd
where {
	VALUES ?spec {<http://meta.icos-cp.eu/resources/cpmeta/atcCo2L2DataObject>}
	?dobj cpmeta:hasObjectSpec ?spec .
	BIND(EXISTS{[] cpmeta:isNextVersionOf ?dobj} AS ?hasNextVersion)
	VALUES ?station {<http://meta.icos-cp.eu/resources/stations/AS_ZSF> <http://meta.icos-cp.eu/resources/stations/AS_HEL> <http://meta.icos-cp.eu/resources/stations/AS_PAL> <http://meta.icos-cp.eu/resources/stations/AS_TRN> <http://meta.icos-cp.eu/resources/stations/AS_GAT> <http://meta.icos-cp.eu/resources/stations/AS_HPB> <http://meta.icos-cp.eu/resources/stations/AS_OXK> <http://meta.icos-cp.eu/resources/stations/AS_IPR> <http://meta.icos-cp.eu/resources/stations/AS_WES> <http://meta.icos-cp.eu/resources/stations/AS_RUN> <http://meta.icos-cp.eu/resources/stations/AS_STE> <http://meta.icos-cp.eu/resources/stations/AS_OPE> <http://meta.icos-cp.eu/resources/stations/AS_CBW> <http://meta.icos-cp.eu/resources/stations/AS_CMN> <http://meta.icos-cp.eu/resources/stations/AS_PUI> <http://meta.icos-cp.eu/resources/stations/AS_BIR> <http://meta.icos-cp.eu/resources/stations/AS_UTO> <http://meta.icos-cp.eu/resources/stations/AS_SNO> <http://meta.icos-cp.eu/resources/stations/AS_IZO> <http://meta.icos-cp.eu/resources/stations/AS_KIT> <http://meta.icos-cp.eu/resources/stations/AS_PRS> <http://meta.icos-cp.eu/resources/stations/AS_SMR> <http://meta.icos-cp.eu/resources/stations/AS_RGL> <http://meta.icos-cp.eu/resources/stations/AS_SAC> <http://meta.icos-cp.eu/resources/stations/AS_ZEP> <http://meta.icos-cp.eu/resources/stations/AS_JUE> <http://meta.icos-cp.eu/resources/stations/AS_LMP> <http://meta.icos-cp.eu/resources/stations/AS_TOH> <http://meta.icos-cp.eu/resources/stations/AS_KRE> <http://meta.icos-cp.eu/resources/stations/AS_SVB> <http://meta.icos-cp.eu/resources/stations/AS_LUT> <http://meta.icos-cp.eu/resources/stations/AS_HTM> <http://meta.icos-cp.eu/resources/stations/AS_SSL> <http://meta.icos-cp.eu/resources/stations/AS_JFJ> <http://meta.icos-cp.eu/resources/stations/AS_PUY> <http://meta.icos-cp.eu/resources/stations/AS_WAO> <http://meta.icos-cp.eu/resources/stations/AS_NOR> <http://meta.icos-cp.eu/resources/stations/AS_LIN>}
			?dobj cpmeta:wasAcquiredBy/prov:wasAssociatedWith ?station .
	?dobj cpmeta:hasSizeInBytes ?size .
?dobj cpmeta:hasName ?fileName .
?dobj cpmeta:wasSubmittedBy/prov:endedAtTime ?submTime .
?dobj cpmeta:hasStartTime | (cpmeta:wasAcquiredBy / prov:startedAtTime) ?timeStart .
?dobj cpmeta:hasEndTime | (cpmeta:wasAcquiredBy / prov:endedAtTime) ?timeEnd .
	FILTER NOT EXISTS {[] cpmeta:isNextVersionOf ?dobj}
	{
		{FILTER NOT EXISTS {?dobj cpmeta:hasVariableName ?varName}}
		UNION
		{
			?dobj cpmeta:hasVariableName ?varName
			FILTER (?varName = "co2")
		}
	}
		{
			{VALUES ?spec {<http://meta.icos-cp.eu/resources/cpmeta/atcCo2L2DataObject>}}
			UNION
			{VALUES ?keyword {"CO2"^^xsd:string}
	?dobj cpmeta:hasKeyword ?keyword}
		}
}
order by desc(?submTime)
```

**Comment by lewisblake** [comment]
We should discuss on Monday the amount of testing needed in this PR since it is a derived class of `ReadMEP` with no additional functionality.

**Review by jgriesfeller — APPROVED** [review]

File: pyaerocom/plugins/icos/reader.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+from __future__ import annotations
+
+import logging
+import re
+from collections import defaultdict
+from collections.abc import Iterable
+from functools import cached_property, lru_cache
+from pathlib import Path
+
+import xarray as xr
```
Comment (avaldebe) [inline pyaerocom/plugins/icos/reader.py:10]: please remove unused modules

File: pyaerocom/plugins/icos/reader.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    # STATION_REGEX = re.compile("icos-co2-nrt-(.*)-.*-.*-.*.nc")
+    STATION_REGEX = re.compile("icos-co2-(.*)-.*-.*.nc")
+
+    DEFAULT_VARS = list(VAR_MAPPING)
+
+    DATASET_NAME = DATA_ID
+
+    PROVIDES_VARIABLES = list(VAR_MAPPING) + list(AUX_FUNS)
+
+    def __init__(self, data_id=None, data_dir=None):
```
Comment (avaldebe) [inline pyaerocom/plugins/icos/reader.py:69]: is i possible to include type annotations?

File: tests/plugins/icos/test_reader.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
+from __future__ import annotations
+
+from pathlib import Path
+
+import pytest
+
+from pyaerocom import const
+from pyaerocom.plugins.mep.reader import ReadMEP
```
Comment (avaldebe) [inline tests/plugins/icos/test_reader.py:8]: please remove unused modules

File: tests/plugins/icos/test_reader.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+from pyaerocom.plugins.icos.reader import ReadICOS
+from tests.conftest import lustre_unavail
+
+try:
+    ICOS_PATH = Path(const.OBSLOCS_UNGRIDDED[const.ICOS_NAME])
+except KeyError:
+    pytest.mark.skip(reason=f"ICOS path not initialised due to non existence in CI")
```
Comment (avaldebe) [inline tests/plugins/icos/test_reader.py:15]: this line sould be somethign like
```python
    pytestmark = pytest.skip("...")
```

File: tests/plugins/icos/test_reader.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+VARS_DEFAULT = {"vmrco2"}
+VARS_PROVIDED = VARS_DEFAULT  # | {} add more if ever needed
+
+STATION_NAMES = ("bir", "gat", "hpb")
+
+
+@lustre_unavail
```
Comment (avaldebe) [inline tests/plugins/icos/test_reader.py:23]: Decorating with `pytest.mark.skipif` only works for test functions.
Decorating a fixture will not mark the tests that use the fixture.

File: tests/plugins/icos/test_reader.py (hunk 4 above)
Comment (avaldebe) [inline tests/plugins/icos/test_reader.py:15]: if all tests need to be decorated with `@lustre_unavail`
all test in the module can be skipped as follows:
``` python
if not const.has_access_lustre:
    pytestmark = pytest.skip("Skipping tests that require access to AEROCOM database on METNo servers")

try:
    ICOS_PATH = Path(const.OBSLOCS_UNGRIDDED[const.ICOS_NAME])
except KeyError:
    pytestmark = pytest.skip("ICOS path not initialised due to non existence in CI")
```

but I suspect that `pytestmark` from the exception is enough

**Review by avaldebe — COMMENTED** [review]
only small changes related to the test fixtures

File: tests/plugins/icos/test_reader.py (hunk 4 above)
Comment (lewisblake) [reply tests/plugins/icos/test_reader.py:15]: #910

File: tests/plugins/icos/test_reader.py (hunk 5 above)
Comment (lewisblake) [reply tests/plugins/icos/test_reader.py:23]: #910

File: tests/plugins/icos/test_reader.py (hunk 3 above)
Comment (lewisblake) [reply tests/plugins/icos/test_reader.py:8]: #910

File: tests/plugins/icos/test_reader.py (hunk 4 above)
Comment (lewisblake) [reply tests/plugins/icos/test_reader.py:15]: #910

File: pyaerocom/plugins/icos/reader.py (hunk 2 above)
Comment (lewisblake) [reply pyaerocom/plugins/icos/reader.py:69]: #910

File: pyaerocom/plugins/icos/reader.py (hunk 1 above)
Comment (lewisblake) [reply pyaerocom/plugins/icos/reader.py:10]: #910

**Pull request merged**